ISSUE-350: Refactor Metadata Preview used variables functionality

aksm opened the pull request:
Resolves #350, #318, and #306.

@DiegoPino I believe this rearrangement bypasses the edge case, but I wasn't able to reproduce those errors so not 100% sure. Even if it does cover those cases, I think this could definitely use further simplification. I'd be glad to open another issue to keep improving on that front later if this looks ok.

DiegoPino commented:
@aksm thanks. I will do a deeper testing tomorrow and let you know how it goes. More soon!

aksm commented:
Hi @DiegoPino, I just updated some PHPDocs and resolved #306 with 74c0ba3 so adding this to AMI can happen right after this is merged.

aksm commented:
@DiegoPino I just pushed a commit with the option to toggle this functionality off.

Also, re: local variables in for loops, as far as I can tell, this doesn't touch them.

DiegoPino commented:
Great. Will do some testing tonight. Thanks 😊 --
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

DiegoPino commented:
@aksm testing now

DiegoPino reviewed (changes requested):
@aksm ok, still issues.

Loaded a IIIF Manifest (v3) with an ADO.

NOTICE: PHP message: TypeError: Drupal\format_strawberryfield\Entity\MetadataDisplayEntity::getTwigVariableNames(): Argument #4 ($set_source) must be of type string, null given, called in /var/www/html/web/modules/contrib/format_strawberryfield/src/Entity/MetadataDisplayEntity.php on line 543 in /var/www/html/web/modules/contrib/format_strawberryfield/src/Entity/MetadataDisplayEntity.php on line 488 #0 /var/www/html/web/modules/contrib/format_strawberryfield/src/Entity/MetadataDisplayEntity.php(543): Drupal\format_strawberryfield\Entity\MetadataDisplayEntity->getTwigVariableNames(Object(Twig\Node\Expression\Binary\RangeBinary), Array, 'pagenumber', NULL)
#1 /var/www/html/web/modules/contrib/format_strawberryfield/src/Entity/MetadataDisplayEntity.php(543): Drupal\format_strawberryfield\Entity\MetadataDisplayEntity->getTwigVariableNames(Object(Twig\Node\ForNode), Array, 'pagenumber', NULL)
#2 /var/www/html/web/modules/contrib/format_strawberryfield/src/Entity/MetadataDisplayEntity.php(543): Drupal\format_strawberryfield\Entity\MetadataDisplayEntity->getTwigVariableNames(Object(Twig\Node\Node), Array, 'document', 'data.as:documen...')

Which means you are not checking if the value passed in the recursive calls to getTwigVariableNames might/not be null..

Also, DCS there are a few (2 I saw) `if(` without a space before.

Give this another run going via the debugger. Maybe the fact that you are depending too much on the Function's default value for arguments makes you be less aware of what you are passing.

DiegoPino commented:
Hint.  $set_source = $parent_path; You sure $parent_path might not be null?

aksm commented:
@DiegoPino, ok, sorry I missed that. One other thing I just noticed while debugging this is the following E_USER_DEPRECATED error:

> Using an "if" condition on "for" tag in "IIIF Presentation API 3.0 Manifest" at line 183 is deprecated since Twig 2.10.0, use a "filter" filter or an "if" condition inside the "for" body instead (if your condition depends on a variable updated inside the loop).

Isn't that something we want to show to the user? Should I check these errors for any mention of "Twig"?

DiegoPino commented:
Question about E_USER_DEPRECATED  you mention, is @aksm: is this avoiding an actual rendering? Means also Search API deprecations (test please with the Search Solr extension we have in archipelago) are permeating/hindering a preview?

DiegoPino commented:
@aksm also: before I test again. Could you please give this another pass/test with real data/ various templates and extensions?
I'm happy to test Monday noon if that means the code is better tested.

aksm commented:
Hi @DiegoPino, I've reduced the recursive Twig variable function by a significant amount but had to add a custom implementation of PHP's array_merge_recursive(). I think this could use further refinement, but I hope it's ok for now. Will do a little further testing with more ADOs and templates and then move onto the display behavior.

DiegoPino commented:
@aksm thanks. let me know when you are done testing and I can give it a pass 😸

aksm commented:
Hi @DiegoPino, I believe this is ready for review now. Thanks!

DiegoPino commented:
@aksm thanks. Will review starting at 2PM EST and let you know how this goes!

DiegoPino commented on src/Entity/MetadataDisplayEntity.php:
```diff
@@ -473,95 +507,57 @@ public function getTwigVariablesUsed() {
+      if ($node instanceof Node) {
+        $add_variables = $this->getTwigVariableNames($node);
+        if (!empty($variables) || !empty($add_variables)) {
+          $variables = static::arrayMergeRecursive($variables, $add_variables);
```
There is excessive recursion here. Basically the arrayMergeRecursive will traverse over and over the same keys (said differently, If the array has N entries, you will traverse N times the last loop, N-1 time the almost last, N-2, etc), when you could really just traverse N times or not at all using https://github.com/esmero/format_strawberryfield/pull/351/files#diff-b5161fcfb4c915fb64182570502fc39348dd9af67772fa46c6e2a7f24f814e3eR553 and passing $variables as a/by reference argument between recursions.

DiegoPino commented on src/Entity/MetadataDisplayEntity.php:
```diff
@@ -473,95 +507,57 @@ public function getTwigVariablesUsed() {
+                      && $node->getAttribute('name') == 'data'
+        ? $node->getAttribute('name')
+        : '';
+      if ($node->hasAttribute('always_defined')) {
```
There is no need to parse anything at all (you can return) if $variable_key is empty (also better approach is to make it NULL instead of ''. That is cheaper and more consistent when checking if.

DiegoPino commented on src/Entity/MetadataDisplayEntity.php:
```diff
@@ -473,95 +507,57 @@ public function getTwigVariablesUsed() {
+              && $nodes->getNode('node')->hasAttribute('name')
+              && $nodes->getNode('node')->getAttribute('name') == 'data'
+      ) {
+        $variable_key = $nodes->getNode('node')->getAttribute('name')
```
Everywhere here $variable_key (first part) will always start with `data` (because of the conditions). I see how you could save  a lot of IFs by grouping conditionals and also setting some values as fixed strings (e.g data)

DiegoPino reviewed (changes requested):
Hi, first pass, there is some performance issues and also some IF/ELSE that are not even needed (you can break before even going there?). If you can add some inline comments explaining what/type of variable parsing is being done by each conditional, that would help me a lot. Like what is a `seq`? When `always_defined` works, etc

aksm commented:
Hi @DiegoPino, I believe I've addressed your above comments with the most recent commits, and I think I'm better able to explain my reasoning this time. Will do a little more testing and double-checking to be sure.

DiegoPino commented:
@aksm still reviewing it but need to pull again after https://github.com/esmero/format_strawberryfield/pull/351/commits/9c059efa31cf84f3aea8dcd5ba48444eea73986c. I need another hour, slow today.

aksm commented on src/Form/MetadataDisplayForm.php:
```diff
@@ -502,11 +502,11 @@ public static function ajaxPreview($form, FormStateInterface $form_state) {
           $output['preview_error'] = $preview_error;
         }
-        if (!$show_render_native || ($show_render_native && $mimetype != 'text/html')) {
+        if ($render && (!$show_render_native || ($show_render_native && $mimetype != 'text/html'))) {
```
When `$render = $entity->renderNative($context);` (line 430 above) throws an exception, e.g. the template can't be rendered because of bad Twig, this generates a warning because `$render` is now undefined. Additionally, since the `$show_json_table` assignment comes after, it also generates the undefined warning in the logs. Should these variables be initialized before?

@DiegoPino thoughts?

aksm replied:
or maybe set $render to false or null in the catch if not set, or add an isset check for both variables?

aksm replied:
Ok, thought it through a little further, and it seems like it makes the most sense to use isset($render) instead of just $render in the conditionals. Will make another commit shortly.

DiegoPino replied:
yep, that seems like easier. Thanks

DiegoPino replied:
Or you can put another try/catch around $render = $entity->renderNative($context); ?

aksm replied:
> Or you can put another try/catch around $render = $entity->renderNative($context); ?

@DiegoPino Is there an advantage to nesting another try/catch? Seems like the outer one catches the same exception either way?

DiegoPino replied:
Not really. I would require to refactor your try catch into two. The issue is because you have a final/ the catch itself (or the final) will have to do the extra work. Just dismiss. the isset() is  enough

aksm commented:
Ok @DiegoPino, sorry about that. I've now tested with all output mime types, multiple output options, multiple templates (new and saved), different exceptions, and even AMI set rows.

DiegoPino commented:
All good. Will have this reviews/tested by tomorrow and we check back on Monday. Have a good weekend

DiegoPino reviewed (changes requested):
@aksm still doing testing here. The actual variable processing looks better (DCS still not perfect) but there is a lot of redundancy in how you are processing that table on the output + DCS. I will keep testing for an hour and then request more changes/have a talk about this during our Team call. Hope we can get it in decent shape for a merge today

aksm commented on src/Entity/MetadataDisplayEntity.php:
```diff
@@ -473,96 +473,50 @@ public function getTwigVariablesUsed() {
    */
-  private function getTwigVariableNames(ModuleNode|Node|BodyNode $nodes, array $all_variables, string $set_var = '', string $set_source = ''): array {
-    $variables = [];
+  private function getTwigVariableNames(
```
Hi @DiegoPino. I've been keeping the coding standards open and referring to them, but I missed that section while jumping between standard PHP conventions and Drupal-specific ones. Will update.

aksm commented on src/Entity/MetadataDisplayEntity.php:
```diff
@@ -473,96 +473,50 @@ public function getTwigVariablesUsed() {
-            $variable_names_flat[] = $name['path'];
-          }
-          $prev_name = $name;
+      // Parse seq to get the values for for/in loops,
```
Sorry, maybe not the clearest example. I meant getting/checking the "name" attribute for "data". Will update with a better example.

DiegoPino commented on src/Form/MetadataDisplayForm.php:
```diff
@@ -287,45 +219,32 @@ public static function buildAjaxPreviewError(string $message) {
-        $data_json[$used_var_path]['used'] = 'Used';
-        $data_json[$used_var_path]['line'] = $used_var_line;
+    $data_json = [];
+    foreach($jsondata as $key=>$value) {
```
DCS > `$key=>$value` should be `$key => $value`

DiegoPino commented on src/Form/MetadataDisplayForm.php:
```diff
@@ -287,45 +219,32 @@ public static function buildAjaxPreviewError(string $message) {
-            $data_json[$used_var_part]['line'] = $used_var_line;
-          }
-        }
+      elseif (is_array($value) && !StrawberryfieldJsonHelper::arrayIsMultiSimple($value)) {
```
Did ::arrayIsMultiSimple do the trick? 😄

DiegoPino commented on src/Form/MetadataDisplayForm.php:
```diff
@@ -287,45 +219,32 @@ public static function buildAjaxPreviewError(string $message) {
+    ksort($data_json);
 
-    $unused_keys = array_keys($unused_vars);
+    $unused_keys = array_keys($data_json);
```
good, no more duplicated json

aksm commented:
Hi @DiegoPino, aside from a little more cleanup and maybe some comments/notes, I think this is in much better shape. If we need to go deeper into the ADO JSON keys, it'll be much easier to add that later instead of starting complex and simplifying (my biggest mistake here). Sorry for the delay, but this is turning out to be a great learning experience for me.

aksm commented:
Hi @DiegoPino, I think the changes based on the recent feedback from you and the team are ready for review. I will do some further looking over while I continue with the documentation.

DiegoPino reviewed (approved):
Code is simpler now and I feel most concerns are addressed. Merging. Thanks @aksm
